src/d3a/models/market/two_sided.py

```diff
@@ -254,18 +228,48 @@ def accept_bid_offer_pair(self, bid, offer, clearing_rate, trade_bid_info, selec
         return bid_trade, trade
 
     def match_offers_bids(self):
-        while len(self._perform_pay_as_bid_matching()) > 0:
-            for bid, offer in self._perform_pay_as_bid_matching():
-                selected_energy = bid.energy if bid.energy < offer.energy else offer.energy
-                original_bid_rate = bid.original_bid_price / bid.energy
-                matched_rate = bid.energy_rate
-
-                trade_bid_info = TradeBidOfferInfo(
-                    original_bid_rate=original_bid_rate,
-                    propagated_bid_rate=bid.price/bid.energy,
-                    original_offer_rate=offer.original_offer_price/offer.energy,
-                    propagated_offer_rate=offer.price/offer.energy,
-                    trade_rate=original_bid_rate)
-
-                self.accept_bid_offer_pair(bid, offer, matched_rate,
-                                           trade_bid_info, selected_energy)
+        pass
+
+    def match_recommendation(self, recommended_list):
+        if recommended_list is None:
+            return
+        for index, recommended_pair in enumerate(recommended_list):
+
+            selected_energy = recommended_pair.bid_energy
+            bid = recommended_pair.bid
+            offer = recommended_pair.offer
+            original_bid_rate = \
+                bid.original_bid_price / bid.energy
+
+            trade_bid_info = TradeBidOfferInfo(
+                original_bid_rate=original_bid_rate,
+                propagated_bid_rate=bid.price/bid.energy,
+                original_offer_rate=offer.original_offer_price/offer.energy,
+                propagated_offer_rate=offer.price/offer.energy,
+                trade_rate=recommended_pair.trade_rate)
+
+            bid_trade, trade = self.accept_bid_offer_pair(
```

Comment:
I don't see where the validation is residing

Reply:
Added in recent commit: https://github.com/gridsingularity/d3a/blob/feature/D3ASIM-3355v1/src/d3a/models/market/__init__.py#L266